### PR TITLE
Feature/allow removal of steps per epoch

### PR DESCRIPTION
**Pull request by mingruimingrui**
fit_generator currently requires steps_per_epoch as a compulsory input. However by adding __len__ to Generator, fit_generator automatically trains the model on all images in Generator object.

### CODE REVIEW

File: keras_retinanet/preprocessing/generator.py [hunk 1]
```diff
@@ -244,3 +244,25 @@ def next(self):
...
+    #
+    # for img, ann in <Generator object>:
+    #     ...
+    #     ...
+    #
+    # A generator is supposed to be callable by next infinitely while an iterator is not
+    # this function should return a copy of this object with an iterator of a fixed length
+    # to avoid complicated situations where fit_generator is unable to call next on this object
+    # def __iter__(self):
+    #     return self
```
Comment (de-vri-es) [inline keras_retinanet/preprocessing/generator.py]: Could you remove the comment about not implementing `__iter__`?  If you want to preserve the reasoning, you can explain why we shouldn't implement `__iter__` in this PR or in the commit message.

**Comment by mingruimingrui** [comment]
My bad, should have done something like don't merge this pull request or something.

`train_generator` should function perfectly as per before in `fit_generator`, with the added benefit that you don't have to define `steps_per_epoch`.

Now if you attempt to call `train_generator` as an iterator, it would throw a `NotImplementedError` since the ability that ability is disabled.

```
train_generator = CocoGenerator(...)

len(train_generator)
> return size of generator

next()
> return next input, target group

for inputs, targets in train_generator:
    ...
    ...
> will throw  NotImplementedError since the ability to call `train_generator` as iterator is disabled
```

File: keras_retinanet/preprocessing/generator.py [hunk 2]
```diff
@@ -244,3 +244,25 @@ def next(self):
             self.group_index = (self.group_index + 1) % len(self.groups)
 
         return self.compute_input_output(group)
+
+    # When left unspecified, steps_per_epoch takes on the value of len(generator)
+    # https://keras.io/models/sequential/#fit_generator
+    # This allows the keras_retinanet.preprocessing.generator.Generator object to be callable by len
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: I like the addition of `__len__`, but the comment isn't required in my opinion.

File: keras_retinanet/preprocessing/generator.py [hunk 3]
```diff
@@ -244,3 +244,25 @@ def next(self):
...
+    def __len__(self):
+        return self.size()
+
+    def __next__(self):
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: This function is already defined ?
https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/preprocessing/generator.py#L234-L235

File: keras_retinanet/preprocessing/generator.py (hunk 1 above)
Comment (hgaiser) [reply keras_retinanet/preprocessing/generator.py]: I agree with @de-vri-es

**Comment by de-vri-es** [comment]
> Personally I am in favour of the latter, but many are in favour of the first. If you plan to add it to the training script, I would prefer to keep the default functionality as is and optionally add an argument --auto-steps or `--steps auto' or something to cover the other case.

I would prefer `--steps auto` even if that is probably slightly more difficult with pythons argparse.

**Comment by de-vri-es** [comment]
> Now if you attempt to call train_generator as an iterator, it would throw a NotImplementedError since the ability that ability is disabled.

That's not new though, is it? Or do these changes disable that before somehow?

**Comment by hgaiser** [comment]
What's the status of this PR? I see there are still some comments that are not addressed.

**Comment by hgaiser** [comment]
Closing due to inactivity.

**Pull request closed without merging**